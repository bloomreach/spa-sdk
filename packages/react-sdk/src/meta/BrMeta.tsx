/*
 * Copyright 2019 Hippo B.V. (http://www.onehippo.com)
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import React from 'react';
import ReactDOM from 'react-dom';
import { MetaComment, Meta, META_POSITION_BEGIN, META_POSITION_END, isMetaComment } from '@bloomreach/spa-sdk';

interface BrMetaProps {
  meta: Meta[];
}

export class BrMeta extends React.Component<BrMetaProps> {
  private comments: Comment[] = [];
  private head?: Element | Text;
  private tailRef = React.createRef<HTMLElement>();

  componentDidMount() {
    this.head = ReactDOM.findDOMNode(this)!;
    this.renderMeta();
  }

  componentDidUpdate() {
    this.head = ReactDOM.findDOMNode(this)!;
    this.removeMeta();
    this.renderMeta();
  }

  componentWillUnmount() {
    this.removeMeta();
  }

  private removeMeta() {
    this.comments.splice(0).forEach(comment => comment.remove());
  }

  private renderMeta() {
    this.props.meta.filter(meta => isMetaComment(meta) && meta.getPosition() === META_POSITION_BEGIN)
      .forEach(this.renderMetaComment.bind(this));
    this.props.meta.filter(meta => isMetaComment(meta) && meta.getPosition() === META_POSITION_END)
      .reverse()
      .forEach(this.renderMetaComment.bind(this));
  }

  private renderMetaComment(meta: MetaComment) {
    const comment = this.head!.ownerDocument!.createComment(meta.getData());
    this.comments.push(comment);

    if (meta.getPosition() === META_POSITION_BEGIN) {
      return void this.head!.parentNode!.insertBefore(comment, this.head!);
    }

    if (!this.tailRef.current!.nextSibling) {
      return void this.tailRef.current!.parentNode!.appendChild(comment);
    }

    this.tailRef.current!.parentNode!.insertBefore(comment, this.tailRef.current!.nextSibling);
  }

  render() {
    return (
      <>
        {this.props.children}
        {this.props.meta.length ? <span style={{ display: 'none' }} ref={this.tailRef} /> : null}
      </>
    );
  }
}
