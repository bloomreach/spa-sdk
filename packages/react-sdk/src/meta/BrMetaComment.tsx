/*
 * Copyright 2019 Hippo B.V. (http://www.onehippo.com)
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import React from 'react';
import { MetaComment } from '@bloomreach/spa-sdk';

interface BrMetaCommentProps {
  meta: MetaComment;
}

export class BrMetaComment extends React.Component<BrMetaCommentProps> {
  private placeholder = React.createRef<HTMLElement>();
  private comment?: Comment;

  componentDidMount() {
    const placeholder = this.placeholder.current!;
    this.comment = placeholder.ownerDocument!.createComment(this.props.meta.getData());
    placeholder.parentNode!.replaceChild(this.comment, placeholder);
  }

  shouldComponentUpdate(nextProps: BrMetaCommentProps) {
    return nextProps.meta.getData() !== this.props.meta.getData();
  }

  componentDidUpdate() {
    this.comment!.replaceData(0, this.comment!.data.length, this.props.meta.getData());
  }

  componentWillUnmount() {
    this.comment!.parentNode!.replaceChild(this.placeholder.current!, this.comment!);
    delete this.comment;
  }

  render() {
    return <span style={{ display: 'none' }} ref={this.placeholder} />;
  }
}
