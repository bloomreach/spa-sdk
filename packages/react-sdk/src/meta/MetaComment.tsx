/*
 * Copyright 2019 Hippo B.V. (http://www.onehippo.com)
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import React from 'react';
import { MetaComment as SpaMetaComment } from '@bloomreach/spa-sdk';

interface MetaCommentProps {
  meta: SpaMetaComment;
}

export class MetaComment extends React.Component<MetaCommentProps> {
  private placeholder = React.createRef<HTMLElement>();
  private comment?: Comment;

  componentDidMount() {
    if (!this.placeholder.current) {
      return;
    }

    const placeholder = this.placeholder.current;
    this.comment = placeholder.ownerDocument!.createComment(this.props.meta.getData());
    placeholder.parentNode!.replaceChild(this.comment, placeholder);
  }

  shouldComponentUpdate(nextProps: MetaCommentProps) {
    return nextProps.meta.getData() !== this.props.meta.getData();
  }

  componentDidUpdate() {
    this.comment!.replaceData(0, this.comment!.data.length, this.props.meta.getData());
  }

  componentWillUnmount() {
    if (!this.comment) {
      return;
    }

    this.comment.remove();
  }

  render() {
    return <span style={{ display: 'none' }} ref={this.placeholder} />;
  }
}
