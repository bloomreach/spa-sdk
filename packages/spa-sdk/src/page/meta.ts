/*
 * Copyright 2019 Hippo B.V. (http://www.onehippo.com)
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

export const META_POSITION_BEGIN = 'begin';
export const META_POSITION_END = 'end';

export type MetaType = string;
export type MetaPosition = typeof META_POSITION_BEGIN | typeof META_POSITION_END;

export interface MetaModel<T extends MetaType = MetaType> {
  data: string;
  type: T;
}

export interface MetaCollectionModel {
  beginNodeSpan?: MetaModel[];
  endNodeSpan?: MetaModel[];
}

/**
 * Meta information describing a part of the page.
 */
export interface Meta<T extends MetaType = MetaType> {
  /**
   * Returns meta-data.
   */
  getData(): string;

  /**
   * Returns meta-data type.
   */
  getType(): T;

  /**
   * Returns meta-data position relative to the related element.
   */
  getPosition(): MetaPosition;
}

export class Meta<T extends MetaType = MetaType> implements Meta {
  constructor(protected model: MetaModel<T>, protected position: MetaPosition) {}

  getData() {
    return this.model.data;
  }

  getType() {
    return this.model.type;
  }

  getPosition() {
    return this.position;
  }
}
