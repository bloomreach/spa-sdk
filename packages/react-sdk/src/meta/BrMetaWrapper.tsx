/*
 * Copyright 2019 Hippo B.V. (http://www.onehippo.com)
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import React from 'react';
import { Meta, META_POSITION_BEGIN, META_POSITION_END } from '@bloomreach/spa-sdk';
import { BrMeta } from './BrMeta';

interface BrMetaWrapperProps {
  meta: Meta[];
}

function renderMeta(meta: Meta[], position: typeof META_POSITION_BEGIN | typeof META_POSITION_END) {
  return meta
    .filter(meta => position === meta.getPosition())
    .map((meta, index) => <BrMeta key={index} meta={meta} />);
}

// tslint:disable-next-line:function-name
export function BrMetaWrapper(props: React.PropsWithChildren<BrMetaWrapperProps>) {
  return (
    <>
      {renderMeta(props.meta, META_POSITION_BEGIN)}
      {props.children}
      {renderMeta(props.meta, META_POSITION_END)}
    </>
  );
}
