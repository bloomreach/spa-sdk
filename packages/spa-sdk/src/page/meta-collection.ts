/*
 * Copyright 2020 Hippo B.V. (http://www.onehippo.com)
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import { Factory } from './factory';
import { MetaModel, Meta, META_POSITION_BEGIN, META_POSITION_END, MetaPosition } from './meta';

/**
 * @hidden
 */
export interface MetaCollectionModel {
  beginNodeSpan?: MetaModel[];
  endNodeSpan?: MetaModel[];
}

/**
 * Collection of the meta-data describing a part of the page.
 * @note The collection extends the built-in Array type for backward compatibility.
 */
export interface MetaCollection extends Array<Meta> {}

export class MetaCollectionImpl extends Array<Meta> implements MetaCollection {
  constructor(model: MetaCollectionModel, factory: Factory<[MetaModel, MetaPosition], Meta>) {
    super(
      ...(model.beginNodeSpan || []).map(model => factory.create(model, META_POSITION_BEGIN)),
      ...(model.endNodeSpan || []).map(model => factory.create(model, META_POSITION_END)),
    );
    Object.setPrototypeOf(this, Object.create(MetaCollectionImpl.prototype));
  }
}
