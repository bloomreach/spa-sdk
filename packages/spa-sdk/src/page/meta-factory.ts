/*
 * Copyright 2019 Hippo B.V. (http://www.onehippo.com)
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import { MetaCollectionModel, MetaModel, MetaPosition, MetaType, Meta, META_POSITION_BEGIN, META_POSITION_END } from './meta';

type MetaBuilder<T extends MetaType> = (model: MetaModel<T>, position: MetaPosition) => Meta<T>;

/**
 * The factory to produce meta-data collection from the page model meta-data.
 */
export class MetaFactory {
  private mapping = new Map<MetaType, MetaBuilder<any>>();

  /**
   * Registers a meta builder for the specified type.
   * @param type The meta type.
   * @param builder The meta builder.
   */
  register<T extends MetaType>(
    type: T,
    builder: MetaBuilder<T>,
  ) {
    this.mapping.set(type, builder);

    return this;
  }

  /**
   * Produces a meta-data collection based on the model.
   * @param meta The meta-data from the page model.
   */
  create(meta?: MetaCollectionModel) {
    return [
      ...(meta && meta.beginNodeSpan || []).map(this.buildMeta.bind(this, META_POSITION_BEGIN)),
      ...(meta && meta.endNodeSpan || []).map(this.buildMeta.bind(this, META_POSITION_END)),
    ];
  }

  private buildMeta(position: MetaPosition, model: MetaModel) {
    const builder = this.mapping.get(model.type);
    if (!builder) {
      throw new Error(`Unsupported meta type: '${model.type}'.`);
    }

    return builder(model, position);
  }
}
