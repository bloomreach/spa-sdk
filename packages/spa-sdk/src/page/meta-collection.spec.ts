/*
 * Copyright 2020 Hippo B.V. (http://www.onehippo.com)
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import { MetaFactory } from './meta-factory';
import { MetaCollectionImpl } from './meta-collection';
import { MetaType, Meta, META_POSITION_BEGIN, META_POSITION_END } from './meta';

describe('MetaCollectionImpl', () => {
  describe('constructor', () => {
    let factory: jest.Mocked<MetaFactory>;

    beforeEach(() => {
      factory = { create: jest.fn() } as unknown as jest.Mocked<MetaFactory>;
    });

    it('should extend a built-in array class', () => {
      const collection = new MetaCollectionImpl({}, factory);

      expect(collection).toBeInstanceOf(Array);
    });

    it('should call a factory to build meta-data', () => {
      new MetaCollectionImpl(
        {
          beginNodeSpan: [
            { data: 'meta1', type: 'type1' as MetaType },
            { data: 'meta2', type: 'type2' as MetaType },
          ],
        },
        factory,
      );

      expect(factory.create).toBeCalledWith({ data: 'meta1', type: 'type1' }, expect.anything());
      expect(factory.create).toBeCalledWith({ data: 'meta2', type: 'type2' }, expect.anything());
    });

    it('should pass a position of the meta', () => {
      new MetaCollectionImpl(
        {
          beginNodeSpan: [{ data: 'meta1', type: 'type1' as MetaType }],
          endNodeSpan: [{ data: 'meta2', type: 'type2' as MetaType }],
        },
        factory,
      );

      expect(factory.create).toBeCalledWith({ data: 'meta1', type: 'type1' }, META_POSITION_BEGIN);
      expect(factory.create).toBeCalledWith({ data: 'meta2', type: 'type2' }, META_POSITION_END);
    });

    it('should hold meta-data items', () => {
      const meta1 = { data: 'meta1' } as unknown as Meta;
      const meta2 = { data: 'meta2' } as unknown as Meta;

      factory.create.mockReturnValueOnce(meta1);
      factory.create.mockReturnValueOnce(meta2);

      const collection = new MetaCollectionImpl(
        {
          beginNodeSpan: [{ data: 'meta1', type: 'type1' as MetaType }],
          endNodeSpan: [{ data: 'meta2', type: 'type2' as MetaType }],
        },
        factory,
      );

      expect(collection).toMatchSnapshot();
    });
  });
});
