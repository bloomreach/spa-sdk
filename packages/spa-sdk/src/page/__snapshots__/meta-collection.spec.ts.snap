// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MetaCollectionImpl constructor should hold meta-data items 1`] = `
MetaCollectionImpl [
  Object {
    "data": "meta1",
  },
  Object {
    "data": "meta2",
  },
]
`;
